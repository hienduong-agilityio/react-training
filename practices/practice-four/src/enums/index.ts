// Index Enum files
export * from './status';
export * from './theme';
