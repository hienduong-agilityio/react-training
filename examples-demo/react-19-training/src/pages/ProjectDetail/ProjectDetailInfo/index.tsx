// Components
import { Avatar, StatusLabel, Tag, CurrencyText, ResourceTag } from '../../../components';

// Enums
import { COLORS } from '../../../enums';

// Interfaces
import type { IProjectItemProps } from '../../../interfaces';

// Helpers
import { getColorFromStatus } from '../../../helpers';

export interface ProjectDetailInfoProps extends Partial<IProjectItemProps> {
  managerName: string;
  managerAvatar?: string;
}

/**
 * Component to render detailed information for a single project.
 *
 * @returns The rendered project detail info.
 */
export const ProjectDetailInfo = ({
  projectName,
  status,
  managerName = 'Unknown',
  managerAvatar = '',
  lastUpdate,
  resources,
  timeline,
  budget
}: ProjectDetailInfoProps) => {
  const color = getColorFromStatus(status as COLORS);

  return (
    <div className='flex items-center p-8 h-full gap-36 space-y-6 flex-row'>
      {/* Manager Information */}
      <div className='flex-shrink-0 w-5/12 h-full flex flex-col items-center'>
        <Avatar name={managerName} src={managerAvatar} customClass='w-3/4 h-full' />
        <p className='text-primary-400 text-lg mt-4 text-center'>{managerName}</p>
      </div>

      {/* Project Overview */}
      <div className='flex-grow lg:w-7/12 h-full space-y-6'>
        <h2 className='text-4xl font-bold text-primary-600'>{projectName}</h2>

        <div className='flex items-center space-x-3 mb-4'>
          <span className='text-2xl font-medium'>Status: </span>
          <StatusLabel color={color} customClasses='text-lg py-[2px] px-3'>
            {status}
          </StatusLabel>
        </div>

        <div className='flex-col'>
          <span className='text-2xl font-medium'>Project Manager</span>
          <p className='text-lg mt-2 text-primary-400'>{managerName}</p>
        </div>

        <div className='flex flex-col gap-2'>
          <span className='text-2xl font-medium'>Last Update</span>
          <p className='text-lg text-primary-400'>{lastUpdate}</p>
        </div>

        <div className='flex flex-col gap-2'>
          <span className='text-2xl font-medium'>Resources</span>
          <ResourceTag resources={resources} />
        </div>

        <div className='flex flex-col gap-2'>
          <span className='text-2xl font-medium'>Time Start</span>
          <Tag customClasses='w-32 p-2 h-6'>{timeline?.timeStart || '-'}</Tag>
        </div>
        <div className='flex flex-col gap-2'>
          <span className='text-2xl font-medium'>Time End</span>
          <Tag customClasses='w-32 p-2 h-6'>{timeline?.timeEnd || '-'}</Tag>
        </div>

        <div className='flex flex-col gap-2'>
          <span className='text-2xl font-medium'>Estimated Cost</span>
          <CurrencyText className='text-lg text-primary-400 font-bold' currency={budget} />
        </div>
      </div>
    </div>
  );
};
