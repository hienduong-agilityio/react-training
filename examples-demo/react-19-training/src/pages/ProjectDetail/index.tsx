// Libraries
import { Navigate, useParams } from 'react-router-dom';

// Components
import { ProjectDetailInfo } from '../../pages/ProjectDetail/ProjectDetailInfo';

// Hooks
import { useProject } from '../../hooks';

// Interfaces
import { IProjectItemProps } from '../../interfaces';
import { JSX } from 'react';

/**
 * Component to render detailed information for a single project.
 *
 * @returns {JSX.Element} The rendered project details.
 */
const ProjectDetail = (): JSX.Element => {
  const { id } = useParams<{ id: string }>();

  // Get project detail by useProject hook
  const project = useProject({ id: id || '' });
  const { data: projectDetail, isQueryProjectDetailPending, error: projectDetailsError } = project;

  // Render loading spinner if data is being fetched
  if (isQueryProjectDetailPending) {
    // return <ProjectDetailSkeleton />;
    return <p>Loading...</p>;
  }

  // If the project is not found, redirect to a 404 page
  if (!projectDetail) {
    return <Navigate to='/404' replace />;
  }

  // Render error message if there is an error in data fetching
  if (projectDetailsError) {
    return <div>Error loading projects: {projectDetailsError.message}</div>;
  }

  const { manager: { managerName = 'Unknown', managerImage = '' } = {} } = projectDetail as IProjectItemProps;

  return <ProjectDetailInfo managerName={managerName} managerAvatar={managerImage} {...projectDetail} />;
};

export default ProjectDetail;
