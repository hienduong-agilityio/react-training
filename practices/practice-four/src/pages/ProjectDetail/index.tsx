// Libraries
import { Navigate, useParams } from 'react-router-dom';

// Components
import { ProjectDetailSkeleton } from '@/components';
import { ProjectDetailInfo } from '@/pages/ProjectDetail/ProjectDetailInfo';

// Hooks
import { usePageSeo, useProject } from '@/hooks';

// Interfaces
import { IProjectItemProps } from '@/interfaces';

const ProjectDetail = (): JSX.Element => {
  const { id } = useParams<{ id: string }>();
  const project = useProject({ id: id ?? '' });
  const { data: projectDetail, isQueryProjectDetailPending, error: projectDetailsError } = project;

  // Extract SEO data if projectDetail is valid
  const seoData =
    !isQueryProjectDetailPending && projectDetail && !Array.isArray(projectDetail)
      ? {
          title: `${projectDetail.projectName}`,
          description: projectDetail.projectName || `Details about project ${projectDetail.projectName}`,
          ogTitle: `${projectDetail.projectName}`,
          ogDescription: projectDetail.projectName || ''
        }
      : null;

  // Always call usePageSeo
  usePageSeo(
    seoData || {
      title: 'Project Details',
      description: 'A powerful tool for managing and tracking projects efficiently.'
    }
  );

  // Handle loading
  if (isQueryProjectDetailPending) return <ProjectDetailSkeleton />;

  // Handle not found
  if (!projectDetail) return <Navigate to='/404' replace />;

  // Handle error
  if (projectDetailsError) {
    return <div>Error loading project: {projectDetailsError.message}</div>;
  }

  const { manager: { managerName = 'Unknown', managerImage = '' } = {} } = projectDetail as IProjectItemProps;

  return <ProjectDetailInfo managerName={managerName} managerAvatar={managerImage} {...projectDetail} />;
};

export default ProjectDetail;
