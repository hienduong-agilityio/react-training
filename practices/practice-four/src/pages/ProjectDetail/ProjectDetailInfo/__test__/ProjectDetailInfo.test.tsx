// Libraries
import { render, screen } from '@testing-library/react';

// Pages
import { ProjectDetailInfoProps, ProjectDetailInfo } from '@/pages/ProjectDetail/ProjectDetailInfo';

// Enums
import { COLORS } from '@/enums';

const defaultProps: ProjectDetailInfoProps = {
  projectName: 'Project Alpha',
  status: 'In Progress',
  managerName: 'John Doe',
  managerAvatar: '',
  lastUpdate: '2024-10-10',
  resources: '5',
  timeline: { timeStart: '2024-01-01', timeEnd: '2024-12-31' },
  budget: 100000
};

describe('ProjectDetailInfo', () => {
  let renderComponent: (props?: Partial<ProjectDetailInfoProps>) => ReturnType<typeof render>;

  beforeEach(() => {
    renderComponent = (props: Partial<ProjectDetailInfoProps> = {}) => {
      return render(<ProjectDetailInfo {...defaultProps} {...props} />);
    };
  });

  it('matches the snapshot for ProjectDetailInfo', () => {
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders project name, status, and manager info', () => {
    renderComponent();

    expect(screen.getByText('Project Alpha')).toBeInTheDocument();
    expect(screen.getByText('Status:')).toBeInTheDocument();
    expect(screen.getByText('In Progress')).toBeInTheDocument();

    const managerNameElements = screen.getAllByText('John Doe');
    expect(managerNameElements).toHaveLength(2);
    expect(managerNameElements[0]).toBeInTheDocument();
    expect(managerNameElements[1]).toBeInTheDocument();
    expect(screen.getByText('2024-10-10')).toBeInTheDocument();
    expect(screen.getByText('5')).toBeInTheDocument();

    const statusLabel = screen.getByText('In Progress');
    expect(statusLabel).toBeInTheDocument();
    expect(statusLabel).toHaveClass('text-lg py-[2px] px-3');
  });

  it('renders fallback values when optional props are missing', () => {
    renderComponent({
      projectName: 'Project Beta',
      status: COLORS.DANGER,
      managerName: 'Jane Smith',
      timeline: {}
    });

    const dashElements = screen.getAllByText('-');
    expect(dashElements).toHaveLength(2);
    expect(dashElements[0]).toBeInTheDocument();
    expect(dashElements[1]).toBeInTheDocument();
  });

  it('renders Avatar with manager name and image', () => {
    renderComponent({ managerAvatar: '/path/to/image' });

    const avatar = screen.getByAltText('John Doe avatar');
    expect(avatar).toBeInTheDocument();
    expect(avatar).toHaveAttribute('src', '/path/to/image');
  });

  it('renders the correct budget in CurrencyText', () => {
    renderComponent({ budget: 50000 });

    expect(screen.getByText(/\$?50000/)).toBeInTheDocument();
  });

  it('renders default timeline if timeline is missing', () => {
    renderComponent({ timeline: {} });

    const dashElements = screen.getAllByText('-');
    expect(dashElements).toHaveLength(2);
    expect(dashElements[0]).toBeInTheDocument();
    expect(dashElements[1]).toBeInTheDocument();
  });
});
