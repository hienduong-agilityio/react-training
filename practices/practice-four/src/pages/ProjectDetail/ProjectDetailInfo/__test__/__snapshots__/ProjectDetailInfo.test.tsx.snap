// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectDetailInfo matches the snapshot for ProjectDetailInfo 1`] = `
<DocumentFragment>
  <div
    class="flex items-center p-8 h-full gap-36 space-y-6 flex-row"
  >
    <div
      class="flex-shrink-0 w-5/12 h-full flex flex-col items-center"
    >
      <div
        class="inline-flex border-2 border-gray-100 items-center justify-center overflow-hidden bg-primary-0 rounded-lg w-3/4 h-full"
      >
        <span
          class="font-bold text-xs text-primary-500"
        >
          JD
        </span>
      </div>
      <p
        class="text-primary-400 text-lg mt-4 text-center"
      >
        John Doe
      </p>
    </div>
    <div
      class="flex-grow lg:w-7/12 h-full space-y-6"
    >
      <h2
        class="text-4xl font-bold text-primary-600"
      >
        Project Alpha
      </h2>
      <div
        class="flex items-center space-x-3 mb-4"
      >
        <span
          class="text-2xl font-medium"
        >
          Status: 
        </span>
        <div
          class="w-max"
        >
          <span
            class="text-lg py-[2px] px-3 inline-flex gap-[6px] rounded font-medium leading-[18px] tracking-[0.36px] flex items-center justify-center color-default"
          >
            <svg
              class="inline-flex color-default"
              fill="none"
              height="6"
              role="img"
              viewBox="0 0 6 6"
              width="6"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="6"
                rx="2"
                width="6"
              />
            </svg>
            In Progress
          </span>
        </div>
      </div>
      <div
        class="flex-col"
      >
        <span
          class="text-2xl font-medium"
        >
          Project Manager
        </span>
        <p
          class="text-lg mt-2 text-primary-400"
        >
          John Doe
        </p>
      </div>
      <div
        class="flex flex-col gap-2"
      >
        <span
          class="text-2xl font-medium"
        >
          Last Update
        </span>
        <p
          class="text-lg text-primary-400"
        >
          2024-10-10
        </p>
      </div>
      <div
        class="flex flex-col gap-2"
      >
        <span
          class="text-2xl font-medium"
        >
          Resources
        </span>
        <div>
          <span
            class="py-[3px] px-2 inline-flex gap-[6px] rounded font-medium leading-[18px] tracking-[0.36px] flex items-center justify-center color-default"
          >
            5
          </span>
        </div>
      </div>
      <div
        class="flex flex-col gap-2"
      >
        <span
          class="text-2xl font-medium"
        >
          Time Start
        </span>
        <span
          class="w-32 p-2 h-6 inline-flex gap-[6px] rounded font-medium leading-[18px] tracking-[0.36px] flex items-center justify-center color-default"
        >
          2024-01-01
        </span>
      </div>
      <div
        class="flex flex-col gap-2"
      >
        <span
          class="text-2xl font-medium"
        >
          Time End
        </span>
        <span
          class="w-32 p-2 h-6 inline-flex gap-[6px] rounded font-medium leading-[18px] tracking-[0.36px] flex items-center justify-center color-default"
        >
          2024-12-31
        </span>
      </div>
      <div
        class="flex flex-col gap-2"
      >
        <span
          class="text-2xl font-medium"
        >
          Estimated Cost
        </span>
        <p
          class="text-lg text-primary-400 font-bold"
        >
          US$ 100000k
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
